README.md
<!--
    This file is part of Invenio.
    Copyright (C) 2015 CERN.

    Invenio is free software; you can redistribute it
    and/or modify it under the terms of the GNU General Public License as
    published by the Free Software Foundation; either version 2 of the
    License, or (at your option) any later version.

    Invenio is distributed in the hope that it will be
    useful, but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
    General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with Invenio; if not, write to the
    Free Software Foundation, Inc., 59 Temple Place, Suite 330, Boston,
    MA 02111-1307, USA.

    In applying this license, CERN does not
    waive the privileges and immunities granted to it by virtue of its status
    as an Intergovernmental Organization or submit itself to any jurisdiction.
-->

# Invenio-Classifier

[![Build Status](https://img.shields.io/travis/inveniosoftware-contrib/invenio-classifier.svg)](https://travis-ci.org/inveniosoftware-contrib/invenio-classifier)
[![Coverage Status](https://img.shields.io/coveralls/inveniosoftware-contrib/invenio-classifier.svg)](https://coveralls.io/r/inveniosoftware-contrib/invenio-classifier)
[![GitHub tag](https://img.shields.io/github/tag/inveniosoftware-contrib/invenio-classifier.svg)](https://github.com/inveniosoftware-contrib/invenio-classifier/releases)
[![PyPI downloads](https://img.shields.io/pypi/dm/invenio-classifier.svg)](https://pypi.python.org/pypi/invenio-classifier)
[![License](https://img.shields.io/github/license/inveniosoftware-contrib/invenio-classifier.svg)](https://github.com/inveniosoftware-contrib/invenio-classifier/blob/master/LICENSE)

Invenio module for record classification.

* Free software: GPLv2 license
* Documentation: https://pythonhosted.org/invenio-classifier

## Features

Classifier automatically extracts keywords from fulltext documents. The
automatic assignment of keywords to textual documents has clear benefits
in the digital library environment as it aids catalogization,
classification and retrieval of documents.

## Keyword extraction is simple

> **Note:** Classifier requires Python [RDFLib](http://rdflib.net/) in order
> to process the RDF/SKOS taxonomy.

In order to extract relevant keywords from a document `fulltext.pdf`
based on a controlled vocabulary `thesaurus.rdf`, you would run
Classifier as follows:

```shell
${INVENIO_WEB_INSTANCE} classifier extract -k thesaurus.rdf -f fulltext.pdf
```

Launching `${INVENIO_WEB_INSTANCE} classifier --help` shows the options available.

As an example, running classifier on document
[nucl-th/0204033](http://cds.cern.ch/record/547024) using the
high-energy physics RDF/SKOS taxonomy (`HEP.rdf`) would yield the
following results (based on the HEP taxonomy from October 10th 2008):

```text
Input file: 0204033.pdf

Author keywords:
Dense matter
Saturation
Unstable nuclei

Composite keywords:
10  nucleus: stability [36, 14]
6  saturation: density [25, 31]
6  energy: symmetry [35, 11]
4  nucleon: density [13, 31]
3  energy: Coulomb [35, 3]
2  energy: density [35, 31]
2  nuclear matter: asymmetry [21, 2]
1  n: matter [54, 36]
1  n: density [54, 31]
1  n: mass [54, 16]

Single keywords:
61  K0
23  equation of state
12  slope
4  mass number
4  nuclide
3  nuclear model
3  mass formula
2  charge distribution
2  elastic scattering
2  binding energy
```

## Thesaurus

Classifier performs an extraction of keywords based on the recurrence
of specific terms, taken from a controlled vocabulary. A controlled
vocabulary is a thesaurus of all the terms that are relevant in a
specific context. When a context is defined by a discipline or branch of
knowledge then the vocabulary is said to be a *subject thesaurus*.
Various existing subject thesauri can be found
[here](http://www.fbi.fh-koeln.de/institut/labor/Bir/thesauri_new/thesen.htm).

A subject thesaurus can be expressed in several different formats.
Different institutions/disciplines have developed different ways of
representing their vocabulary systems. The taxonomy used by classifier
is expressed in RDF/SKOS. It allows not only to list keywords but to
specify relations between the keywords and alternative ways to represent
the same keyword.

```xml
<Concept rdf:about="http://cern.ch/thesauri/HEP.rdf#scalar">
 <composite rdf:resource="http://cern.ch/thesauri/HEP.rdf#Composite.fieldtheoryscalar"/>
 <prefLabel xml:lang="en">scalar</prefLabel>
 <note xml:lang="en">nostandalone</note>
</Concept>

<Concept rdf:about="http://cern.ch/thesauri/HEP.rdf#fieldtheory">
 <composite rdf:resource="http://cern.ch/thesauri/HEP.rdf#Composite.fieldtheoryscalar"/>
 <prefLabel xml:lang="en">field theory</prefLabel>
 <altLabel xml:lang="en">QFT</altLabel>
 <hiddenLabel xml:lang="en">/field theor\w*/</hiddenLabel>
 <note xml:lang="en">nostandalone</note>
</Concept>

<Concept rdf:about="http://cern.ch/thesauri/HEP.rdf#Composite.fieldtheoryscalar">
 <compositeOf rdf:resource="http://cern.ch/thesauri/HEP.rdf#scalar"/>
 <compositeOf rdf:resource="http://cern.ch/thesauri/HEP.rdf#fieldtheory"/>
 <prefLabel xml:lang="en">field theory: scalar</prefLabel>
 <altLabel xml:lang="en">scalar field</altLabel>
</Concept>
```

In RDF/SKOS, every keyword is wrapped around a *concept* which
encapsulates the full semantics and hierarchical status of a term -
including synonyms, alternative forms, broader concepts, notes and so on
- rather than just a plain keyword.

The specification of the SKOS language and [various
manuals](http://www.w3.org/TR/2005/WD-swbp-thesaurus-pubguide-20050517/)
that aid the building of a semantic thesaurus can be found at the [SKOS
W3C
website](http://www.w3.org/TR/2005/WD-swbp-skos-core-guide-20051102/).
Furthermore, Classifier can function on top of an extended version of
SKOS, which includes special elements such as key chains, composite
keywords and special annotations.

## Keyword extraction

Classifier computes the keywords of a fulltext document based on the
frequency of thesaurus terms in it. In other words, it calculates how
many times a thesaurus keyword (and its alternative and hidden labels,
defined in the taxonomy) appears in a text and it ranks the results.
Unlike other similar systems, Classifier does not use any machine
learning or AI methodologies - a just plain phrase matching using
[regular expressions](http://en.wikipedia.org/wiki/Regex): it
exploits the conformation and richness of the thesaurus to produce
accurate results. It is then clear that Classifier performs best on top
of rich, well-structured, subject thesauri expressed in the RDF/SKOS
language.

## Running tests

Use the provided Dockerfile to run the tests with all dependencies installed.
From the repository root, build the image and run the suite:

```shell
docker build -t invenio-classifier-tests .
docker run --rm invenio-classifier-tests poetry run pytest
```

The test run also prints a coverage report, including missing lines.

Happy hacking and thanks for flying Invenio-Classifier.

Inspirehep Development Team
&nbsp;&nbsp;Email: admin@inspirehep.net
&nbsp;&nbsp;Twitter: http://twitter.com/inveniosoftware
&nbsp;&nbsp;GitHub: https://github.com/inveniosoftware-contrib/invenio-classifier
&nbsp;&nbsp;URL: http://inveniosoftware.org


CHANGES.rst
..
    This file is part of Invenio.
    Copyright (C) 2015, 2018 CERN.

    Invenio is free software; you can redistribute it
    and/or modify it under the terms of the GNU General Public License as
    published by the Free Software Foundation; either version 2 of the
    License, or (at your option) any later version.

    Invenio is distributed in the hope that it will be
    useful, but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
    General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with Invenio; if not, write to the
    Free Software Foundation, Inc., 59 Temple Place, Suite 330, Boston,
    MA 02111-1307, USA.

    In applying this license, CERN does not
    waive the privileges and immunities granted to it by virtue of its status
    as an Intergovernmental Organization or submit itself to any jurisdiction.

Changes
=======

Version 1.3.2 (release 2018-03-12)
----------------------------------

Bug fixes
~~~~~~~~~

- Prevent the "unprintable KeyError" problem from happening.

Version 1.3.1 (release 2017-06-23)
----------------------------------

Bug fixes
~~~~~~~~~

- Reverse sort core keywords.

Version 1.3.0 (release 2017-06-22)
----------------------------------

Incompatible changes
~~~~~~~~~~~~~~~~~~~~

- Exports list of objects rather than list of tuples, to help ES
  directly indexing the content.

Version 1.2.0 (release 2017-06-21)
----------------------------------

Incompatible changes
~~~~~~~~~~~~~~~~~~~~

- Do not use keywords as dictionary keys, rather as elements in a list.

Version 1.1.2 (release 2017-05-22)
----------------------------------

Bug fixes
~~~~~~~~~

- Supports '·' author separator
- Support utf8 author-keywords

Version 1.1.1 (release 2017-05-19)
----------------------------------

Bug fixes
~~~~~~~~~

- Enforce utf8 also for non PDF files in extractor.

Version 1.1.0 (release 2017-05-17)
----------------------------------

Incompatible changes
~~~~~~~~~~~~~~~~~~~~

- Changes dict export format for author keywords, into an improved and semantic
  way.
- Renames keys in dict export to be lower case and separated by `_`.

Bug fixes
~~~~~~~~~

- Drop trailing dots in author keywords.

Version 1.0.1 (release 2017-01-11)
----------------------------------

Incompatible changes
~~~~~~~~~~~~~~~~~~~~

- Changes module to be compatible with Invenio 3.

Bug fixes
~~~~~~~~~

- Fixes a crash when trying to discover a taxonomy when
  CLASSIFIER_WORKDIR is set to None.
- Updates minimum dependencies of Invenio packages to newer versions.
- Removes a bug in bibclassify_keyword_analyzer.py. If a combination
  is found via a synonym or regexp it is no longer thrown away just
  because the components of the combination are not found in the text.
- Adds missing `invenio_base` dependency.

Version 0.1.0 (release 2015-08-19)
----------------------------------

- Initial public release.


invenio_classifier/config.py
# -*- coding: utf-8 -*-
#
# This file is part of Invenio.
# Copyright (C) 2007, 2008, 2009, 2010, 2011, 2013, 2014, 2015, 2016 CERN.
#
# Invenio is free software; you can redistribute it and/or
# modify it under the terms of the GNU General Public License as
# published by the Free Software Foundation; either version 2 of the
# License, or (at your option) any later version.
#
# Invenio is distributed in the hope that it will be useful, but
# WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
# General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with Invenio; if not, write to the Free Software Foundation, Inc.,
# 59 Temple Place, Suite 330, Boston, MA 02111-1307, USA.

"""Classifier can be configured in different ways."""

from __future__ import unicode_literals

import re

try:
    from shutil import which
except ImportError:
    # CPython <3.3
    from distutils.spawn import find_executable as which


CLASSIFIER_WORKDIR = None
"""Path to directory for classifier related files, default: instance_path."""

CLASSIFIER_PATH_GFILE = which("file")
"""Path to file executable."""

CLASSIFIER_PATH_PDFTOTEXT = which("pdftotext")
"""Path to pdf2text executable."""

CLASSIFIER_DEFAULT_OUTPUT_NUMBER = 20
"""Number of keywords that are printed by default.
This limits single keywords, composite keywords, and acronyms - not author
keywords."""

CLASSIFIER_RECORD_KEYWORD_FIELD = '6531_'
"""The main marc xml field where to find/save the keywords."""

CLASSIFIER_RECORD_KEYWORD_OTHER_FIELDS = ['6950_']
"""Other fields to take from the marc xml when generating list of keywords."""

CLASSIFIER_RECORD_KEYWORD_AUTHOR_FIELD = ''
"""Where to save author supplied keywords."""

CLASSIFIER_RECORD_KEYWORD_ACRONYM_FIELD = ''
"""Where to save extracted acronyms."""

CLASSIFIER_PARTIAL_TEXT_PERCENTAGES = ((0, 20), (40, 60))
"""Marks the part of the fulltext to keep when running a partial match.
Each tuple contains the start and end percentages of a section."""

CLASSIFIER_INVARIABLE_WORDS = (
    "any", "big", "chi", "der", "eta", "few",
    "low", "new", "non", "off", "one", "out",
    "phi", "psi", "rho", "tau",
    "two", "van", "von", "hard", "weak", "four",
    "anti", "zero", "sinh",
    "open", "high", "data", "dark", "free",
    "flux", "fine", "final", "heavy",
    "strange"
)
"""If the keyword belongs in 'INVARIABLE_WORDS', we return it without
any change."""

CLASSIFIER_EXCEPTIONS = {
    "aluminium": r"alumini?um",
    "aluminum": r"alumini?um",
    "analysis": r"analy[sz]is",
    "analyzis": r"analy[sz]is",
    "behavior": r"behaviou?rs?",
    "behaviour": r"behaviou?rs?",
    "color": r"colou?rs?",
    "colour": r"colou?rs?",
    "deflexion": r"defle(x|ct)ions?",
    "flavor": r"flavou?rs?",
    "flavour": r"flavou?rs?",
    "gas": r"gas(s?es)?",
    "lens": r"lens(es)?",
    "matrix": r"matri(x(es)?|ces)",
    "muon": r"muons?",
    "neutrino": r"neutrinos?",
    "reflexion": r"refle(x|ct)ions?",
    "ring": r"rings?",
    "status": r"status(es)?",
    "string": r"strings?",
    "sum": r"sums?",
    "vertex": r"vert(ex(es)?|ices)",
    "vortex": r"vort(ex(es)?|ices)",
}
"""If the keyword is found in 'EXCEPTIONS', we return its attached regular
expression."""

CLASSIFIER_UNCHANGE_REGULAR_EXPRESSIONS = (
    re.compile("[^e]ed$"),
    re.compile("ics?$"),
    re.compile("[io]s$"),
    re.compile("ium$"),
    re.compile("less$"),
    re.compile("ous$"),
)
"""If the keyword is matched by a regular expression of
'UNCHANGE_REGULAR_EXPRESSIONS', we return the keyword without any
change.
"""

CLASSIFIER_GENERAL_REGULAR_EXPRESSIONS = (
    (re.compile("ional"), r"ional(ly)?"),
    (re.compile("([ae])n(ce|t)$"), r"\1n(t|ces?)"),
    (re.compile("og(ue)?$"), r"og(ue)?s?"),
    (re.compile("([^aeiouyc])(re|er)$"), r"\1(er|re)s?"),
    (re.compile("([aeiouy])[sz]ation$"), r"\1[zs]ations?"),
    (re.compile("([aeiouy])[sz]ation$"), r"\1[zs]ations?"),
    (re.compile("([^aeiou])(y|ies)$"), r"\1(y|ies)"),
    (re.compile("o$"), r"o(e?s)?"),
    (re.compile("(x|sh|ch|ss)$"), r"\1(es)?"),
    (re.compile("f$"), r"(f|ves)"),
    (re.compile("ung$"), r"ung(en)?"),
    (re.compile("([^aiouy])s$"), r"\1s?"),
    (re.compile("([^o])us$"), r"\1(i|us(es)?)"),
    (re.compile("um$"), r"(a|ums?)"),
)
"""The classifier creates a regular expression for each label found in the
ontology.

At last, we perform the sub method of Python's re module using the
first element of the tuple as the regex and the second element as the
replacement string.

Regular expressions found here have been originally based on
Wikipedia's page on English plural.
[http://en.wikipedia.org/wiki/English_plural]
"""

CLASSIFIER_SEPARATORS = {
    " ": r"[\s\n-]",
    "-": r"[\s\n-]?",
    "/": r"[/\s]?",
    "(": r"\s?\(",
    "*": r"[*\s]?",
    "- ": r"\s?\-\s",
    "+ ": r"\s?\+\s",
}
"""The transformation for the non-alpha characters that can be found between
two words."""

CLASSIFIER_SYMBOLS = {
    "'": r"\s?\'",
}
"""The punctuation that can be found at the end of a word."""

CLASSIFIER_WORD_WRAP = "[^\w-]%s[^\w-]"
"""Regular expression to wrap words."""

CLASSIFIER_VALID_SEPARATORS = (
    "of", "of a", "of an", "of the", "of this", "of one", "of two", "of three",
    "of new", "of other", "of many", "of both", "of these", "of each", "is",
    "the"
)
"""When searching for composite keywords, we allow two keywords separated by
one of the component of 'VALID_SEPARATORS' to form a composite keyword. These
separators contain also the punctuation."""


CLASSIFIER_AUTHOR_KW_START = \
    re.compile(r"(?i)key[ -]*words?[a-z ]*[.:] *")

CLASSIFIER_AUTHOR_KW_END = (
    re.compile(r"\n"),
    re.compile(r"\.\W"),
    re.compile(r"\sPACS"),
    re.compile(r"(?i)1[. ]*introduction\W"),
    re.compile(r"(?i)mathematics subject classification\W"),
)

CLASSIFIER_AUTHOR_KW_SEPARATION = re.compile(" ?; ?| ?, ?| ?- | ?· ")
